File: sach/de3331.cpp
//DE3331.CPP - Tran Huu Nam - hsgtin.vn - 2024
#include <bits/stdc++.h>
using namespace std;
int n;
long long s, a[201];
bool d[500001]={};
vector<int> c;
int main() {
    cin>>n>>s;
    for (int i=1; i<=n; i++) cin>>a[i];
    d[0]=true;
    int t=0;
    for (int i=1; i<=n; i++) {
        for (int x=t; x>=0; x--)
            if (d[x] && x+a[i]<=s) d[x+a[i]]=true;
        t+=a[i];
    }
    if (!d[s]) {
        cout<<-1;
        return 0;
    }
    for (int i=n; i>=1; i--)
        if (s>=a[i] && d[s-a[i]]) {
            c.push_back(a[i]);
            s-=a[i];
        }
    cout<<1<<endl;
    for (int i=c.size()-1; i>=0; i--)
        cout<<c[i]<<" ";
    return 0;
}

File: sach/de3331.py
//DE3331.PY - Tran Huu Nam - hsgtin.vn - 2024
import sys
n = 0
s = 0
a = [0] * 201
d = [False] * 500001
c = []
def main():
    global n, s, a, d, c
    input = sys.stdin.read
    data = input().split()
    index = 0
    n = int(data[index])
    index += 1
    s = int(data[index])
    index += 1
    for i in range(1, n + 1):
        a[i] = int(data[index])
        index += 1
    d[0] = True
    t = 0
    for i in range(1, n + 1):
        for x in range(t, -1, -1):
            if d[x] and x + a[i] <= s:
                d[x + a[i]] = True
        t += a[i]
    if not d[s]:
        print(-1)
        return
    for i in range(n, 0, -1):
        if s >= a[i] and d[s - a[i]]:
            c.append(a[i])
            s -= a[i]
    print(1)
    for i in range(len(c) - 1, -1, -1):
        print(c[i], end=" ")
if __name__ == "__main__":
    main()

